Google.Solutions.CloudIap.IapClient/IapClient.cs

```diff
@@ -0,0 +1,177 @@
+// specific language governing permissions and limitations
+// under the License.
+//
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Util.Store;
+using Google.Solutions.Compute;
+using Google.Solutions.Compute.Iap;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Solutions.CloudIap.IapClient
+{
+    internal class IapClient
+    {
+        private enum ExitCode : int
+        {
+            Success = 0,
+            InvalidParameters = 1,
+            UnhandledException = 2
+        }
+
+        private const string ProgramName = "IapClient";
+
+        public VmInstanceReference Instance { get; }
+        public ushort Port { get; }
+
+        public IapClient(VmInstanceReference instance, ushort port)
+        {
+            this.Instance = instance;
+            this.Port = port;
+        }
+
+        public async Task RunAsync(CancellationToken token)
+        {
+            ICredential credential = await GoogleWebAuthorizationBroker.AuthorizeAsync(
+                new ClientSecrets()
+                {
+                    ClientId = "78381520511-4fu6ve6b49kknk3dkdnpudoi0tivq6jn.apps.googleusercontent.com",
+                    ClientSecret = "dRgZl1efp_JKcUqQusuaVIrS"
+                },
+                new[] { IapTunnelingEndpoint.RequiredScope },
+                Environment.UserName,
+                CancellationToken.None,
+                new FileDataStore(GetType().Namespace),
+                new LocalServerCodeReceiver(Resources.AuthorizationSuccessful));
+
+            var iapEndpoint = new IapTunnelingEndpoint(
+                    credential,
+                    this.Instance,
+                    this.Port,
+                    IapTunnelingEndpoint.DefaultNetworkInterface);
+
+            // Probe connection.
+            using (var stream = new SshRelayStream(iapEndpoint))
+            {
+                await stream.TestConnectionAsync(TimeSpan.FromSeconds(2)).ConfigureAwait(false);
+            }
+
+            // Start listener to enable clients to connect.
+            var listener = SshRelayListener.CreateLocalListener(iapEndpoint);
+
+            Console.WriteLine($"Forwarding localhost:{listener.LocalPort} "+
+                $"to {this.Instance}:{this.Port}...");
+
+            listener.ClientConnected += (sender, args) =>
+            {
+                using (new CustomConsoleColor(ConsoleColor.Yellow))
+                {
+                    Console.WriteLine($"Client {args.Client} connected");
+                }
+            };
+            listener.ClientDisconnected += (sender, args) =>
+            {
+                using (new CustomConsoleColor(ConsoleColor.Yellow))
+                {
+                    Console.WriteLine($"Client {args.Client} disconnected");
+                }
+            };
+            listener.ConnectionFailed += (sender, exception) =>
+            {
+                DisplayError(exception.Exception);
```

Comment:
i guess this'd be what users would see from the comment above?

Reply:
Yes